File: server/core/src/main/scala/org/infinispan/server/core/transport/StatsChannelHandler.scala
package org.infinispan.server.core.transport

import java.net.SocketAddress

import io.netty.buffer.ByteBuf
import io.netty.channel._

/**
 * Input/Output ChannelHandler to keep statistics
 *
 * @author gustavonalle
 * @since 7.1
 */
trait StatsChannelHandler extends ChannelInboundHandlerAdapter with ChannelOutboundHandler {

   val transport: NettyTransport

   def bind(ctx: ChannelHandlerContext, localAddress: SocketAddress, promise: ChannelPromise): Unit = ctx.bind(localAddress, promise)

   def connect(ctx: ChannelHandlerContext, remoteAddress: SocketAddress, localAddress: SocketAddress, promise: ChannelPromise): Unit = ctx.connect(remoteAddress, localAddress, promise)

   def disconnect(ctx: ChannelHandlerContext, promise: ChannelPromise): Unit = ctx.disconnect(promise)

   def close(ctx: ChannelHandlerContext, promise: ChannelPromise): Unit = ctx.close(promise)

   def deregister(ctx: ChannelHandlerContext, promise: ChannelPromise): Unit = ctx.deregister(promise)

   def read(ctx: ChannelHandlerContext): Unit = ctx.read()

   override def channelRead(ctx: ChannelHandlerContext, msg: scala.Any): Unit = {
      transport.updateTotalBytesRead(msg.asInstanceOf[ByteBuf].readableBytes())
      super.channelRead(ctx, msg)
   }

   override def channelActive(ctx: ChannelHandlerContext) {
      transport.acceptedChannels.add(ctx.channel)
      super.channelActive(ctx)
   }

   override def write(ctx: ChannelHandlerContext, msg: scala.Any, promise: ChannelPromise): Unit = {
      val readable = msg.asInstanceOf[ByteBuf].readableBytes()
      ctx.write(msg, promise.addListener(new ChannelFutureListener {
         def operationComplete(future: ChannelFuture): Unit = {
            if (future.isSuccess) {
               transport.updateTotalBytesWritten(readable)
            }
         }
      }))
   }

   def flush(ctx: ChannelHandlerContext): Unit = ctx.flush()

}

File: server/core/src/main/scala/org/infinispan/server/core/transport/StatsChannelHandler.java
package org.infinispan.server.core.transport;

import io.netty.buffer.ByteBuf;
import io.netty.buffer.ByteBufHolder;
import io.netty.channel.ChannelDuplexHandler;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelPromise;
import io.netty.util.Attribute;
import io.netty.util.AttributeKey;

import java.time.Instant;

/**
 * Input/Output ChannelHandler to keep statistics
 *
 * @author gustavonalle
 * @author wburns
 * @since 7.1
 */
public class StatsChannelHandler extends ChannelDuplexHandler {
   private final NettyTransport transport;
   public static AttributeKey<Integer> bytesRead = AttributeKey.valueOf("__bytesRead");
   public static AttributeKey<Instant> startInstant = AttributeKey.valueOf("__startInstant");

   public StatsChannelHandler(NettyTransport transport) {
      this.transport = transport;
   }

   @Override
   public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
      int readable = getByteSize(msg);

      Attribute<Integer> count = ctx.channel().attr(bytesRead);
      count.set(readable);
      Attribute<Instant> start = ctx.channel().attr(startInstant);
      start.set(Instant.now());

      transport.updateTotalBytesRead(readable);
      super.channelRead(ctx, msg);
   }

   @Override
   public void channelActive(ChannelHandlerContext ctx) throws Exception {
      transport.acceptedChannels().add(ctx.channel());
      super.channelActive(ctx);
   }

   @Override
   public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
      transport.updateTotalBytesWritten(getByteSize(msg));
      super.write(ctx, msg, promise);
   }

   int getByteSize(Object msg) {
      if (msg instanceof ByteBuf) {
         return ((ByteBuf) msg).readableBytes();
      } else if (msg instanceof ByteBufHolder) {
         return ((ByteBufHolder) msg).content().readableBytes();
      } else {
         return -1;
      }
   }
}
